### PR TITLE
auto disable sandbox for interactive shell

### LINKED ISSUE #38652: Sandbox & Autojump
Seems to be a sandbox thing. I've never hit it before. Reproduces across multiple formulae with the same method.

Essentially, with ` autojump ` installed, start to build something, interrupt the build, go into the shell, do ` cd [hit-your-tab-key-here] ` => ` cd [the-first-two-letters-of-a-folder-the-first-tab-revealed] ` => ` hit your tab key to let it autocomplete the folder name and hit enter `

My config is zsh, with oh-my-zsh and autojump, but it happens with autojump and system ` sh ` as well, at least.

Not particularly easy to describe, so:

CC: @xu-cheng

### PR DESCRIPTION
**Pull request by xu-cheng**
Closes #38652.

cc @DomT4 @Homebrew/owners

### CODE REVIEW

**Comment by DomT4** [comment]
Presume if I run ` brew install --debug example --sandbox ` the ` --sandbox ` argument would get ignored in favour of the ` debug ` one?

I'd prefer a more selective choice personally, in terms of being able to build debugged with the sandbox on by default, and just turn it off with something like ` brew install --debug example --no-sandbox ` - But the code potentially gets more convoluted the more choice we add.

I'm happy here if everyone else is, Thanks! :+1:

**Comment by mikemcquaid** [comment]
> Presume if I run brew install --debug example --sandbox the --sandbox argument would get ignored in favour of the debug one?

I think it'd be good in cases like this if there's either a message printed or a conflicting arguments error.

**Comment by mikemcquaid** [comment]
Otherwise :+1:

**Comment by xu-cheng** [comment]
Updated with a warning message.

**Comment by mikemcquaid** [comment]
Might want to check this warning message isn't printed more than once (I've been bitten by that in the past)

**Comment by xu-cheng** [comment]
Yeah, it will print twice in `brew install` (one for build one for postinstall). And I fail to find a simple way to prevent it happening. `ARGV.sandbox?` is invoked inside a fork. I found I cannot store a module variable in such case.

File: Library/Homebrew/extend/ARGV.rb [hunk 1]
```diff
@@ -99,7 +99,12 @@ def homebrew_developer?
   end
 
   def sandbox?
-    include?("--sandbox") || !ENV["HOMEBREW_SANDBOX"].nil?
+    enable = include?("--sandbox") || !ENV["HOMEBREW_SANDBOX"].nil?
+    autodisable = interactive? || debug?
+    if enable && autodisable
+      opoo "Sandbox will be auto disabled in debug or interactive mode."
```
Comment (mikemcquaid) [inline Library/Homebrew/extend/ARGV.rb]: `The sandbox cannot be used in debug or interactive mode.`

**Comment by mikemcquaid** [comment]
It's probably worth pulling the message printing into just e.g. `formula_installer` and out of `ARGV`.

**Comment by xu-cheng** [comment]
> It's probably worth pulling the message printing into just e.g. formula_installer and out of ARGV.

It's also used in `test`.  And frankly, I think it's OK to let this warning repeat multiple times since every sandbox is independent.

File: Library/Homebrew/extend/ARGV.rb [hunk 2]
```diff
@@ -99,7 +99,12 @@ def homebrew_developer?
   end
 
   def sandbox?
-    include?("--sandbox") || !ENV["HOMEBREW_SANDBOX"].nil?
+    enable = include?("--sandbox") || !ENV["HOMEBREW_SANDBOX"].nil?
+    autodisable = interactive? || debug?
+    if enable && autodisable
+      opoo "The sandbox cannot be used in debug or interactive mode."
+    end
+    enable && !autodisable
```
Comment (mikemcquaid) [inline Library/Homebrew/extend/ARGV.rb]: Looking at the other functions in this file I think probably all of this code should be moved to a utility function elsewhere and the method left as it was.

Comment (xu-cheng) [inline Library/Homebrew/extend/ARGV.rb]: Where should I call such utility function?

Comment (mikemcquaid) [inline Library/Homebrew/extend/ARGV.rb]: Wherever you're calling `ARGV.sandbox?` and where `debug` or `interactive` are also called.

Comment (xu-cheng) [inline Library/Homebrew/extend/ARGV.rb]: How about put them into `Sandbox::available?`? or as a separate method like `Sandbox::auto disable?`

Comment (mikemcquaid) [inline Library/Homebrew/extend/ARGV.rb]: Either sounds good :+1:

Comment (xu-cheng) [inline Library/Homebrew/extend/ARGV.rb]: Updated. I think `Sandbox::available?` should use as a feature checker(like things in `brew config`). Therefore, I added  `Sandbox::auto_disable?`.

File: Library/Homebrew/sandbox.rb [hunk 3]
```diff
@@ -8,6 +8,14 @@ def self.available?
     OS.mac? && File.executable?(SANDBOX_EXEC)
   end
 
+  # there're times sandbox is inconvenience
```
Comment (mikemcquaid) [inline Library/Homebrew/sandbox.rb]: `there are times the sandbox cannot be used`

File: Library/Homebrew/sandbox.rb [hunk 4]
```diff
@@ -8,6 +8,14 @@ def self.available?
...
+  def self.auto_disable?
+    if (auto_disable = ARGV.interactive? || ARGV.debug?)
+      opoo "The sandbox cannot be used in debug or interactive mode."
```
Comment (mikemcquaid) [inline Library/Homebrew/sandbox.rb]: Still feels potentially weird to have a `?` method which is normally a boolean getter print a message without any arguments? Could maybe take an argument so it's only printed once?

File: Library/Homebrew/sandbox.rb (hunk 3 above)
Comment (xu-cheng) [inline Library/Homebrew/sandbox.rb]: Maybe `should not be used`? As actually, we can use it, it's just inconvenience.

**Comment by mikemcquaid** [comment]
I'm OK with merging this if we can work out a way to avoid the message being printed multiple times.

**Comment by xu-cheng** [comment]
Unfortunately, I haven't find a way to avoid the duplicated message as this method is called inside a fork.

**Comment by mikemcquaid** [comment]
Worst case could just save some state inside the method? If that's not possible, explicitly require an argument to be passed and only call with that argument the first time.

**Comment by xu-cheng** [comment]
@mikemcquaid I re-implement this to make it only show warning message once.

**Comment by mikemcquaid** [comment]
@xu-cheng :+1: 

I'd love us to start integrating the sandbox soon. Maybe we could add it as a DSL method or something (like `env :std`) as we slowly migrate core to work with it?

**Comment by xu-cheng** [comment]
I'm kinda aginst the idea of a DSL to disable sandbox. Although, sandbox currently doesn't improve security in a great deal. From a security aspect, formula should not have the ability to turn off a security protocol. That power should belong to the ender users.

**Comment by mikemcquaid** [comment]
Long-term, I agree. In the short-term it's the only way I can see us getting the sandbox enabled globally. Up to you, though, it's your feature! I'm not vastly concerned about security but it does help us enforce e.g. formulae installing into others' kegs which causes weird, hard-to-debug bugs.

**Pull request closed without merging**